Argo/Globals/StandardTypes.swift

```diff
@@ -2,53 +2,65 @@ import Foundation
 import Runes
 
 extension String: JSONDecodable {
-  public static func decode(j: JSON) -> String? {
+  public static func fromJSON(j: JSON) -> Parser<String> {
     switch j {
-    case let .String(s): return s
-    default: return .None
+    case let .String(s): return pure(s)
+    default: return typeMismatch("String", j)
     }
   }
 }
 
 extension Int: JSONDecodable {
-  public static func decode(j: JSON) -> Int? {
+  public static func fromJSON(j: JSON) -> Parser<Int> {
     switch j {
-    case let .Number(n): return n as Int
-    default: return .None
+    case let .Number(n): return pure(n as Int)
+    default: return typeMismatch("Int", j)
     }
   }
 }
 
 extension Double: JSONDecodable {
-  public static func decode(j: JSON) -> Double? {
+  public static func fromJSON(j: JSON) -> Parser<Double> {
     switch j {
-    case let .Number(n): return n as Double
-    default: return .None
+    case let .Number(n): return pure(n as Double)
+    default: return typeMismatch("Double", j)
     }
   }
 }
 
 extension Bool: JSONDecodable {
-  public static func decode(j: JSON) -> Bool? {
+  public static func fromJSON(j: JSON) -> Parser<Bool> {
     switch j {
-    case let .Number(n): return n as Bool
-    default: return .None
+    case let .Number(n): return pure(n as Bool)
+    default: return typeMismatch("Bool", j)
     }
   }
 }
 
 extension Float: JSONDecodable {
-  public static func decode(j: JSON) -> Float? {
+  public static func fromJSON(j: JSON) -> Parser<Float> {
     switch j {
-    case let .Number(n): return n as Float
-    default: return .None
+    case let .Number(n): return pure(n as Float)
+    default: return typeMismatch("Float", j)
     }
   }
 }
 
-public func decodeArray<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> [A]? {
+public func decodeArray<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> Parser<[A]> {
   switch value {
-  case let .Array(a): return sequence({ A.decode($0) } <^> a)
-  default: return .None
+  case let .Array(a): return sequence({ A.fromJSON($0) } <^> a)
+  default: return typeMismatch("Array", value)
   }
 }
+
+public func decodeObject(value: JSON) -> Parser<Dictionary<String, JSON>> {
+  switch value {
+  case let .Object(o): return pure(o)
+  default: return typeMismatch("Object", value)
+  }
+}
+
+private func typeMismatch<T>(type: String, object: JSON) -> Parser<T> {
```

Comment:
I do think this is useful enough for users that want to write their own parsers that it should probably be `public`. But I still think it should be a static method on `Parser` instead of a free function.